arlo_server/rounds.py

```diff
@@ -0,0 +1,150 @@
+from flask import jsonify, request
+from jsonschema import validate
+from werkzeug.exceptions import BadRequest, Conflict
+import uuid, datetime
+
+from arlo_server import app, db
+from arlo_server.models import (
+    Round,
+    RoundContest,
+    Election,
+    SampledBallot,
+    SampledBallotDraw,
+)
+from arlo_server.auth import with_election_access, UserType
+from arlo_server.sample_sizes import sample_size_options
+from util.isoformat import isoformat
+from audits import sampler
+
+
+CREATE_ROUND_REQUEST_SCHEMA = {
+    "type": "object",
+    "properties": {
+        "roundNum": {"type": "integer", "minimum": 1,},
+        "sampleSize": {"type": "integer", "minimum": 1,},
+    },
+    "additionalProperties": False,
+    "required": ["roundNum"],
+}
+
+
+def serialize_round(r: Round) -> dict:
+    return {
+        "id": r.id,
+        "roundNum": r.round_num,
+        "startedAt": isoformat(r.started_at),
+        "endedAt": isoformat(r.ended_at),
+    }
+
+
+# Raises if invalid
+def validate_round(r: dict, election: Election):
+    validate(r, CREATE_ROUND_REQUEST_SCHEMA)
+
+    rounds = sorted(election.rounds, key=lambda r: r.round_num, reverse=True)
+    current_round = next(iter(rounds), None)
+    next_round_num = current_round.round_num + 1 if current_round else 1
+    if r["roundNum"] != next_round_num:
+        raise BadRequest(f"The next round should be round number {next_round_num}")
+
+    if current_round and not current_round.ended_at:
+        raise Conflict(f"The current round is not complete")
+
+    if r["roundNum"] == 1 and "sampleSize" not in r:
+        raise BadRequest(f"Sample size is required for round 1")
+
+
+def sample_ballots(election: Election, round: Round, sample_size: int):
+    # For now, we only support one targeted contest
+    contest = next(c for c in election.contests if c.is_targeted)
+
+    # Compute the total number of ballot samples in all rounds leading up to
+    # this one. Note that this corresponds to the number of SampledBallotDraws,
+    # not SampledBallots.
+    num_previously_sampled = (
+        SampledBallotDraw.query.join(Round).filter_by(election_id=election.id).count()
+    )
+
+    # Create the pool of ballots to sample (aka manifest) by combining the
+    # manifests from every jurisdiction in the contest's universe.
+    # Audits must be deterministic and repeatable for the same real world
+    # inputs. So the sampler expects the same input for the same real world
+    # data. Thus, we use the jurisdiction and batch names (deterministic real
+    # world ids) instead of the jurisdiction and batch ids (non-deterministic
+    # uuids that we generate for each audit).
+    manifest = {
+        (jurisdiction.name, batch.name): batch.num_ballots
```

Comment:
It seems sufficient for now to me, since each batch in the contest universe should be uniquely identified by jurisdiction+batch.

I guess the only caveat I could think of (which is maybe what you're getting at) is that if we need to change this later to add more fields, then audits run before that change might not sample the same ballots as identical audits run after that change. Is that an issue? I can't really think of any way to work around it if that's the case.

Reply:
I think future changes to this will be for batch manifests, where we have (jurisdiction, batch name, number of ballots). Or CVRs too. I'm not super sure this is very important right now, I just wanted to flag it so we could think about it a little more. 

Another thing to worry about is when there's a collision between jurisdiction and batch names, e.g. if Ann Arbor City and Ann Arbor Township both omit their full designations and they share a precinct name. So you'd get (Ann Arbor, PS1) twice, with no way to disambiguate. That can be handled by assertions or something like that during data entry, I suppose.

Reply:
Your point about collisions seems very important. I think we should definitely enforce the uniqueness of jurisdiction names within an election and batch names within a jurisdiction. We can do this at the db level. cc @benadida to confirm

Reply:
@jonahkagan I agree on requiring unique jurisdiction names within an audit, and unique batch names within a manifest. Worth asking @mcchilders if she foresees any issue, but my sense is the complexity goes up a LOT if we can't make this assumption.

Reply:
Confirmed with @mcchilders. Added task: #389